modules/composer-workflowoperation/src/main/java/org/opencastproject/workflow/handler/composer/MultiEncodeWorkflowOperationHandler.java

```diff
@@ -0,0 +1,545 @@
+   *
+   * Source Tracks are selected by (Flavor AND Tag) if they are both provided
+   *
+   * There can be multiple sources and flavors to create more than one source tracks. In the workflow, A semi-colon ";"
+   * is used to separate the independent operations.
+   *
+   * The independent operations can be either all share the same set of properties or all have different sets of
+   * properties. For example, There are two sets of source flavors: * "presenter/* ; presentation/*", one source tag,
+   * eg: "preview", and two sets of encoding profiles, eg: "mp4,flv ; mp4,hdtv" then there are two concurrent
+   * operations: the first one is all "presenter" tracks tagged "preview" will be encoded with "mp4" and "flv". The
+   * second one is all "presentation" tracks tagged "preview" encoded with "mp4" and "hdtv"
+   *
+   */
+  private List<ElementProfileTagFlavor> getSrcSelector(String[] sourceFlavors, String[] sourceTags,
+          String[] targetFlavors, String[] targetTags, String[] profiles) throws WorkflowOperationException {
+    int n = 0;
+    List<ElementProfileTagFlavor> elementSelectors = new ArrayList<ElementProfileTagFlavor>();
+    if (sourceTags == null && sourceFlavors == null)
+      throw new WorkflowOperationException("No source tags or Flavor");
+    if (profiles == null)
+      throw new WorkflowOperationException("Missing profiles");
+    if (sourceTags != null) { // If source tags are used to select tracks
+      // If use source and target tags, there should be the same number of them or all map into one target
+      if (targetTags != null && (targetTags.length != 1 && sourceTags.length != targetTags.length))
+        throw new WorkflowOperationException("number of source tags " + sourceTags.length
+                + " does not match number of target tags " + targetTags.length + " (must be the same or one target)");
+      // There should be the same number of source tags or profile groups or all use same group of profiles
+      if (profiles.length != 1 && sourceTags.length != profiles.length) {
+        throw new WorkflowOperationException(
+                "number of source tags segments " + sourceTags.length + " does not match number of profiles segments "
+                        + profiles.length + " (must be the same or one profile)");
+      }
+      // If use source tags and source flavors, there should be the same number of them or one
+      if (sourceFlavors != null && (sourceTags.length != 1 && sourceFlavors.length != 1)
+              && sourceFlavors.length != sourceTags.length) {
+        throw new WorkflowOperationException("number of source tags segments " + sourceTags.length
+                + " does not match number of source Flavor segments " + sourceFlavors.length
+                + " (must be the same or one)");
+      }
+      n = sourceTags.length; // at least this many tracks
+    }
+    if (sourceFlavors != null) { // If flavors are used to select tracks
```

Comment:
(reading top down): Maybe it would be worth to explicitly address all three cases: Only source-tags, only source-flavors or both source-tags and source-flavors as the intersection of tracks selected by both source-tags and source-tracks seems a special case